File: src/main/scala/5.longest-palindromic-substring.scala
/**
  * 动态规划实现最长回文串
  */
object Q5DP {
  def longestPalindrome(s: String): String = {
    val len = s.length
    var res = ""
    if (len != 0) {
      // dp缓存数组
      val dp = Array.ofDim[Boolean](len, len)
      // until 右边不闭合
      for (r <- 0 until len) {
        // to 右边闭合
        for (l <- 0 to r) {
          if (s.charAt(l) == s.charAt(r) && (r - l <= 2 || dp(l + 1)(r - 1))) {
            dp(l)(r) = true
            val curLen = r - l + 1
            if (curLen > res.length) {
              res = s.substring(l, r + 1)
            }
          }
        }
      }
    }
    res
  }
}

/**
  * fp风格的暴力法, 无法通过测试
  * 参考:
  * https://stackoverflow.com/questions/15253591/scala-laziness-how-to-get-rid-of-return-statement-in-this-palindrome-search
  */
object Q5BruteForce {
  def longestPalindrome(s: String): String = {
    (for {
      substrSize <- s.length to 1 by -1
      // sliding生成多个窗口, 对切割好的窗口再迭代返回
      substr <- s.sliding(substrSize)
      if substr == substr.reverse
      // 因为是倒序遍历, 所以前面的肯定是更长的字串
    } yield substr).headOption.getOrElse("")
  }
}

File: src/main/java/5.longest-palindromic-substring.java
/**
 * 第5题用dp重做
 * 
 * <pre>
 * dp[i][j] 表示从i to j 这段路径是否为回文
 * {@code dp[i][j] = dp[i + 1][j - 1] && s[i] == s[j]}
 *
 * 循环i从大到小, j从小到大, 举例:
 * 因为 dp[0][4] 依赖 dp[1][3], 所以i必须从大往小算, j必须从小到大算
 * </pre>
 */
class Q5DP {
  public String longestPalindrome(String s) {
    String ret = "";
    int n = s.length();
    boolean[][] dp = new boolean[n][n];
    for (int i = n - 1; i >= 0; i--) {
      for (int j = i; j < n; j++) {
        // 排除首尾不相登的肯定不是回文
        if (s.charAt(i) != s.charAt(j)) {
          continue;
        }
        // 首尾相等, 并且中间只有一位或者中间也是回文
        if (j - i < 2 || dp[i + 1][j - 1]) {
          // 距离在2以内肯定是回文, 单字或双字
          dp[i][j] = true;
        }
        if (dp[i][j] && j - i + 1 > ret.length())
          ret = s.substring(i, j + 1);
      }
    }
    return ret;
  }
}
